### PR TITLE
Added character screen

### PR DESCRIPTION
**Pull request by nicolaurenti**
# Modified Main screen UI

In this PR I modified the main screen to add a title, a description, a button and a gif 

## Unit Test

image

## Screenshot

Nexus S          |  Pixel 3A
:-------------------------:|:-------------------------:
Nexus S | Pixel 3a

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -28,4 +32,4 @@
             android:resource="@array/preloaded_fonts" />
     </application>
 
-</manifest>
+</manifest>
```
Comment (nicolaurenti) [inline app/src/main/AndroidManifest.xml]: Add newline

File: app/src/main/java/com/marvelapp/presentation/viewmodel/CharacterViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    data class CharactersData(
+        val state: CharactersState,
+        val data: List<CharacterEntity> = listOf()
+    )
+
+    enum class CharactersState {
+        WAITING_CHARACTERS,
+        SHOW_CHARACTERS
+    }
+}
```
Comment (nicolaurenti) [inline app/src/main/java/com/marvelapp/presentation/viewmodel/CharacterViewModel.kt]: Same

File: app/src/main/res/drawable/rounded_background.xml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shape="oval">
+    <corners android:radius="500dp" />
+    <solid
+        android:color="@color/white" />
+</shape>
```
Comment (nicolaurenti) [inline app/src/main/res/drawable/rounded_background.xml]: .

File: app/src/main/java/com/marvelapp/data/CharacterEntity.kt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package com.marvelapp.data
+
+data class CharacterEntity(
+    val name: String,
+    val image: String? = null,
```
Comment (SartiMau) [inline app/src/main/java/com/marvelapp/data/CharacterEntity.kt]: 1) remove this unnecessary comma
2) it is really necessary that this be nullable and with a default value?

File: app/src/main/java/com/marvelapp/presentation/activity/CharacterActivity.kt [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        setContentView(binding.root)
+        lifecycleScope.launchWhenStarted {
+            viewModel.state.collectLatest { this@CharacterActivity.updateUI(it) }
+        }
+        viewModel.showCharacters()
+    }
+
+    private fun updateUI(state: CharacterViewModel.CharactersData) {
+        when (state.state) {
+            WAITING_CHARACTERS -> {} // DO NOTHING
```
Comment (SartiMau) [inline app/src/main/java/com/marvelapp/presentation/activity/CharacterActivity.kt]: why do you have this DO NOTHING?

File: app/src/main/java/com/marvelapp/presentation/activity/MainMenuActivity.kt [hunk 6]
```diff
@@ -31,16 +32,23 @@ class MainMenuActivity : AppCompatActivity() {
         lifecycleScope.launchWhenStarted {
             viewModel.state.collectLatest { this@MainMenuActivity.updateUI(it) }
         }
+        binding.mainMenuButton.setOnClickListener { viewModel.buttonPressed() }
     }
 
     private fun updateUI(state: MainMenuViewModel.MainMenuStates) {
         when (state) {
             INIT -> {
                 // DO NOTHING
```
Comment (SartiMau) [inline app/src/main/java/com/marvelapp/presentation/activity/MainMenuActivity.kt]: why is this here?

File: app/src/main/java/com/marvelapp/presentation/activity/MainMenuActivity.kt [hunk 7]
```diff
@@ -31,16 +32,23 @@ class MainMenuActivity : AppCompatActivity() {
...
             }
+            GO_TO_CHARACTER_SCREEN -> {
+                goToCharacterScreen()
+            }
         }
     }
 
+    private fun goToCharacterScreen() {
+        startActivity(CharacterActivity.getIntent(this))
```
Comment (SartiMau) [inline app/src/main/java/com/marvelapp/presentation/activity/MainMenuActivity.kt]: you can avoid creating this method and put this line in the correspondent option of the when statement

File: app/src/main/java/com/marvelapp/presentation/adapter/CharacterAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    override fun onBindViewHolder(holder: CharacterViewHolder, position: Int) {
+        holder.bind(characters[position])
+    }
+
+    override fun getItemCount(): Int = characters.size
+
+    class CharacterViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+        private val binding = ItemCharacterBinding.bind(itemView)
+        fun bind(item: CharacterEntity) {
+            val requestOptions = RequestOptions().placeholder(R.drawable.file_not_found).error(R.drawable.file_not_found)
```
Comment (SartiMau) [inline app/src/main/java/com/marvelapp/presentation/adapter/CharacterAdapter.kt]: avoid creating this variable and waste memory. as this is only used one time, you can create this in the line 30

File: app/src/main/java/com/marvelapp/presentation/activity/CharacterActivity.kt (hunk 5 above)
Comment (SartiMau) [reply app/src/main/java/com/marvelapp/presentation/activity/CharacterActivity.kt]: as it is for the default value needed by the stateFlow, you can rename the state to INIT and call `viewModel.showCharacters()` here

File: app/src/main/java/com/marvelapp/presentation/viewmodel/CharacterViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class CharacterViewModel @Inject constructor() : ViewModel() {
+
+    private val charactersListTest = listOf(CharacterEntity("Batman"), CharacterEntity("Superman"))
```
Comment (SartiMau) [inline app/src/main/java/com/marvelapp/presentation/viewmodel/CharacterViewModel.kt]: why is this needed?

File: app/src/main/java/com/marvelapp/presentation/activity/MainMenuActivity.kt (hunk 6 above)
Comment (SartiMau) [reply app/src/main/java/com/marvelapp/presentation/activity/MainMenuActivity.kt]: as it is for the default value needed by the stateFlow, you can use it to fall Glide here

File: app/src/main/java/com/marvelapp/presentation/viewmodel/MainMenuViewModel.kt [hunk 10]
```diff
@@ -4,13 +4,18 @@ import androidx.lifecycle.ViewModel
...
     private val _state = MutableStateFlow(MainMenuStates.INIT)
     val state: StateFlow<MainMenuStates> = _state.asStateFlow()
 
+    fun buttonPressed() =
+        _state.update { MainMenuStates.GO_TO_CHARACTER_SCREEN }
+
     enum class MainMenuStates {
-        INIT
+        INIT,
+        GO_TO_CHARACTER_SCREEN,
```
Comment (SartiMau) [inline app/src/main/java/com/marvelapp/presentation/viewmodel/MainMenuViewModel.kt]: remove this unnecessary comma

File: app/src/main/res/values/themes.xml [hunk 11]
```diff
@@ -13,4 +13,13 @@
         <item name="android:statusBarColor" tools:targetApi="l">?attr/colorPrimaryVariant</item>
         <!-- Customize your theme here. -->
     </style>
+
+    <style name="NoActionBar">
+        <item name="windowActionBar">false</item>
+        <item name="windowNoTitle">true</item>
+    </style>
+
+    <style name="AppBarOverlay" parent="ThemeOverlay.AppCompat.Dark.ActionBar" />
```
Comment (SartiMau) [inline app/src/main/res/values/themes.xml]: why are these styles needed?

File: app/src/test/java/com/marvelapp/presentation/viewmodel/CharacterViewModelTest.kt [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+import androidx.arch.core.executor.testing.InstantTaskExecutorRule
+import com.marvelapp.data.CharacterEntity
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import kotlinx.coroutines.test.resetMain
+import kotlinx.coroutines.test.runTest
+import kotlinx.coroutines.test.setMain
+import org.junit.*
```
Comment (SartiMau) [inline app/src/test/java/com/marvelapp/presentation/viewmodel/CharacterViewModelTest.kt]: do not use wildcard imports

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/test/java/com/marvelapp/presentation/viewmodel/CharacterViewModelTest.kt (hunk 12 above)
Comment (nicolaurenti) [reply app/src/test/java/com/marvelapp/presentation/viewmodel/CharacterViewModelTest.kt]: Solved

File: app/src/main/res/values/themes.xml (hunk 11 above)
Comment (nicolaurenti) [reply app/src/main/res/values/themes.xml]: Solved

File: app/src/main/java/com/marvelapp/presentation/viewmodel/MainMenuViewModel.kt (hunk 10 above)
Comment (nicolaurenti) [reply app/src/main/java/com/marvelapp/presentation/viewmodel/MainMenuViewModel.kt]: Solved

File: app/src/main/java/com/marvelapp/presentation/activity/MainMenuActivity.kt (hunk 7 above)
Comment (nicolaurenti) [reply app/src/main/java/com/marvelapp/presentation/activity/MainMenuActivity.kt]: Solved

File: app/src/main/java/com/marvelapp/presentation/adapter/CharacterAdapter.kt (hunk 8 above)
Comment (nicolaurenti) [reply app/src/main/java/com/marvelapp/presentation/adapter/CharacterAdapter.kt]: Solved

File: app/src/main/java/com/marvelapp/presentation/viewmodel/CharacterViewModel.kt (hunk 2 above)
Comment (nicolaurenti) [reply app/src/main/java/com/marvelapp/presentation/viewmodel/CharacterViewModel.kt]: Solved

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (nicolaurenti) [reply app/src/main/AndroidManifest.xml]: Solved

File: app/src/main/res/drawable/rounded_background.xml (hunk 3 above)
Comment (nicolaurenti) [reply app/src/main/res/drawable/rounded_background.xml]: Solved

File: app/src/main/java/com/marvelapp/presentation/activity/MainMenuActivity.kt (hunk 6 above)
Comment (SartiMau) [reply app/src/main/java/com/marvelapp/presentation/activity/MainMenuActivity.kt]: this still here

File: app/src/main/java/com/marvelapp/presentation/viewmodel/CharacterViewModel.kt (hunk 9 above)
Comment (SartiMau) [reply app/src/main/java/com/marvelapp/presentation/viewmodel/CharacterViewModel.kt]: this still here

File: app/src/main/java/com/marvelapp/data/CharacterEntity.kt (hunk 4 above)
Comment (nicolaurenti) [reply app/src/main/java/com/marvelapp/data/CharacterEntity.kt]: Solved

File: app/src/main/java/com/marvelapp/presentation/activity/CharacterActivity.kt (hunk 5 above)
Comment (nicolaurenti) [reply app/src/main/java/com/marvelapp/presentation/activity/CharacterActivity.kt]: Solved

File: app/src/main/java/com/marvelapp/presentation/viewmodel/CharacterViewModel.kt (hunk 9 above)
Comment (nicolaurenti) [reply app/src/main/java/com/marvelapp/presentation/viewmodel/CharacterViewModel.kt:17]: Solved

File: app/src/main/java/com/marvelapp/presentation/activity/MainMenuActivity.kt (hunk 6 above)
Comment (nicolaurenti) [reply app/src/main/java/com/marvelapp/presentation/activity/MainMenuActivity.kt]: Solved

File: app/src/main/java/com/marvelapp/presentation/viewmodel/CharacterViewModel.kt (hunk 9 above)
Comment (nicolaurenti) [reply app/src/main/java/com/marvelapp/presentation/viewmodel/CharacterViewModel.kt]: This list is a mock data

File: app/src/main/java/com/marvelapp/presentation/activity/CharacterActivity.kt [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+        setContentView(binding.root)
+        lifecycleScope.launchWhenStarted {
+            viewModel.state.collectLatest { this@CharacterActivity.updateUI(it) }
+        }
+        viewModel.showCharacters()
+    }
+
+    private fun updateUI(state: CharacterViewModel.CharactersData) {
+        when (state.state) {
+            INIT -> showCharactersList(state.data)
```
Comment (SartiMau) [inline app/src/main/java/com/marvelapp/presentation/activity/CharacterActivity.kt]: here you should call `        viewModel.showCharacters()`
and in the SHOW_CHARACTERS state call this

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/marvelapp/presentation/activity/CharacterActivity.kt (hunk 13 above)
Comment (nicolaurenti) [reply app/src/main/java/com/marvelapp/presentation/activity/CharacterActivity.kt]: Solved

**Review by SartiMau — APPROVED** [review]
